Fixed multiple ordering bugs involving unsafe a-temporal abstract values

sb98052 opened the pull request:
Release notes: Fixed multiple ordering bugs

Fixes #2327 and #2406. The discussion around this set of issues can be found in #2327. To summarize, sometimes the serializer's CSE algorithm would relocate simplified values out of the path along which they were simplified. This relocation is unsafe for values that can cause side effects and ones that can throw. Otherwise, it may be wasteful but safe.

This PR helps improves the situation by making all potentially unsafe abstract values temporal. The cases I found came out of a code audit. Calls to `createFromTemplate` should have been covered comprehensively. Most unsafe operations still don't accept abstract values and throw introspection errors, so they are not a problem yet.

sebmarkbage commented:
I don’t understand the criteria decided upon. How are not everything temporal according to this definition?

I believe that almost every expression in JavaScript can throw if given invalid input. So when should I mark something as temporal?

sb98052 commented:
I think a more complete statement of the criterion would be "the value produced by the template can throw under a condition that is not precluded by the assumptions under which the template is used." This is the root of the specific bug that this PR tries to knock off, since such dynamic conditions are sensitive to simplifications by the simplifier. 

The general problem of formalizing the constraints under which something can be atemporal and balancing those constraints with efficiency is still open, and this PR doesn't help with that.

sebmarkbage commented on src/methods/get.js:
```diff
@@ -344,7 +344,10 @@ export function OrdinaryGetPartial(
 ): Value {
   if (Receiver instanceof AbstractValue && Receiver.getType() === StringValue && P === "length") {
-    return AbstractValue.createFromTemplate(realm, lengthTemplate, NumberValue, [Receiver], lengthTemplateSrc);
+    return AbstractValue.createTemporalFromTemplate(realm, lengthTemplate, NumberValue, [Receiver], {
```
This is a simple example. Given that the type of the Receiver is a string, this can never throw or have side-effects. It can only error if this is fed with the wrong type of input arg. That needs to be fixed.

However, once we fix that, how will I know that this can be turned atemporal again? Not everything has this property. So I have to go through them all one-by-one and think hard if this can really throw or not.

hermanventer replied:
Simple: if Receiver is temporal, the get call must be temporal. Otherwise it is a-temporal.

sebmarkbage commented:
Another approach could be to require all abstract values to be derived using the generator but add a flag like `isTemporal` to distinguish things that could be reordered vs things that can't.

hermanventer commented on src/evaluators/BinaryExpression.js:
```diff
@@ -199,7 +199,14 @@ export function computeBinary(
+        // "in" and "instanceof" can throw and therefore must result in temporals
+        if (result instanceof AbstractValue && (op === "in" || op === "instanceof")) {
+          invariant(realm.generator !== undefined);
+          result = realm.generator.deriveAbstract(result.types, result.values, result.args, result.operationDescriptor);
```
Only AbstractValue methods should ever call deriveAbstract directly.

hermanventer commented on src/evaluators/BinaryExpression.js:
```diff
@@ -199,7 +199,14 @@ export function computeBinary(
-        return AbstractValue.createFromBinaryOp(realm, op, lval, rval, loc);
+        let result = AbstractValue.createFromBinaryOp(realm, op, lval, rval, loc);
+
+        // "in" and "instanceof" can throw and therefore must result in temporals
```
If getPureBinaryOperationResultType does not throw, it should not be the case that the binary operation may throw or have a side effect.

sb98052 commented:
> > "the value produced by the template can throw under a condition that is not precluded by the assumptions under which the template is used"
> 
> This seems like circular reasoning to me. This can only be used under the conditions we use it.

More like "This can only throw under the conditions in which it is expected to throw" which is not circular. It's the throwing under certain conditions that is precluded, not the conditions themselves. But my phrasing didn't help in making this clear.

That said, the distinction is irrelevant if the precondition for a value to be atemporal is that it never throw at runtime.

I do see your concern though - my approach makes some values atemporal in a conditional setting, but at the cost of making many others temporal, even in an unconditional setting. As a side effect of my change, calls to `String.length` and such would never get lifted, and that's bad.

I'll try to come up with an approach that strikes a better balance, given these parameters.

hermanventer commented on src/intrinsics/ecma262/JSON.js:
```diff
@@ -317,6 +317,8 @@ const JSONParse = buildExpressionTemplate(JSONParseStr);
 function InternalJSONClone(realm: Realm, val: Value): Value {
   if (val instanceof AbstractValue) {
     if (val instanceof AbstractObjectValue) {
+      // These may need to use the temporal variants of createFromTemplate
```
In particular, if val is temporal, then the derived values must be temporal as well. The reason being that val is known to be an object (and thus not null or undefined) at this temporal point, but may be null or undefined at a different temporal point. Since stringify and parse are not total when their arguments are null or undefined, they cannot become a-temporal when their arguments are not also a-temporal.

hermanventer commented on src/intrinsics/ecma262/JSON.js:
```diff
@@ -317,6 +317,8 @@ const JSONParse = buildExpressionTemplate(JSONParseStr);
     if (val instanceof AbstractObjectValue) {
+      // These may need to use the temporal variants of createFromTemplate
+      // See #2327
       let strVal = AbstractValue.createFromTemplate(realm, JSONStringify, StringValue, [val], JSONStringifyStr);
```
Are you fixing things or just adding a todo comment? If the former, this is not complete, if the latter then please add TODO to the comment along with an issue number.

hermanventer commented on src/intrinsics/ecma262/StringPrototype.js:
```diff
@@ -577,7 +577,10 @@ export default function(realm: Realm, obj: ObjectValue): ObjectValue {
 
     if (O instanceof AbstractValue && O.getType() === StringValue) {
-      return AbstractValue.createFromTemplate(realm, sliceTemplate, StringValue, [O, start, end], sliceTemplateSrc);
+      return AbstractValue.createTemporalFromTemplate(realm, sliceTemplate, StringValue, [O, start, end], {
```
You only need to to do this if O is temporal.

hermanventer commented on src/intrinsics/ecma262/StringPrototype.js:
```diff
@@ -611,13 +614,10 @@ export default function(realm: Realm, obj: ObjectValue): ObjectValue {
-        [O, separator, limit],
-        splitTemplateSrc
-      );
+      return AbstractValue.createTemporalFromTemplate(realm, splitTemplate, ObjectValue, [O, separator, limit], {
```
Check if O is temporal.

sb98052 commented:
Thanks. Herman's insight strikes the balance we want, and also refines the criterion. In general, a value must be temporal if it is dynamic (can throw, side effects). The assumption that something is not dynamic may itself rely on dynamic parameters - and those should be temporal. Conversely, if a parameter is temporal, then a value that relies on such an assumption become temporal too.

I've updated my implementation to do this. Also, I have added inverted tests to make sure that the operations I have touched get lifted in the unconditional case.

*This is still work in progress because turning on the currently commented out invariant in `createAbstractConcreteValue` breaks a few things. Also, some tests are breaking that I need to look into.*

hermanventer commented on src/evaluators/BinaryExpression.js:
```diff
@@ -194,11 +194,28 @@ export function computeBinary(
+        // may rely on a dynamic condition. We express this condition as an implication,
+        // rval is temporal => this operation is temporal
+        // See #2327
+        if ((op === "in" || op === "instanceof") && (rvalIsAbstract && rval.isTemporal()))
```
Rather put this condition into a helper function named something like "operationMustBeTemporal".

sb98052 replied:
I've deferred this to the refactoring in #2489.

hermanventer commented on src/evaluators/BinaryExpression.js:
```diff
@@ -194,11 +194,28 @@ export function computeBinary(
+            createOperationDescriptor("BINARY_EXPRESSION", { op }),
+            { isPure: true, skipInvariant: true }
+          );
+        else return AbstractValue.createFromBinaryOp(realm, op, lval, rval, loc);
```
I'm still a little bit disturbed by this. If one of the arguments is temporal, then moving the operation to point where the current path conditions do not apply may be safe, but it seems hardly desirable.

In essence, any operation over temporal values should result in a temporal value, as a rule, except for a few specific cases that are well documented.

sb98052 replied:
I'm going to enforce this protocol in a follow up PR that addresses #2489. Doing so will involve auditing uses of all of the abstract value factory methods.

hermanventer commented on src/intrinsics/ecma262/JSON.js:
```diff
@@ -317,8 +317,17 @@ const JSONParse = buildExpressionTemplate(JSONParseStr);
+      let obVal;
+
+      // If val is temporal, then this operation is also temporal. See #2327.
+      if (val.isTemporal()) {
```
This logic seems the inverse of what you intend. This seems to call for an additional test case.

sb98052 replied:
It was inverted. On the other hand, it is not obvious to me that InternalJSONClone can create the particular conditions that this PR is trying to fix. So I have dropped this check for now, and will audit in a follow up request.

hermanventer commented on src/values/AbstractValue.js:
```diff
@@ -877,6 +886,11 @@ export default class AbstractValue extends Value {
+
+    // Abstract values in an abstract concrete union are temporal, as they are predicated
+    // on the conditions that preclude the concrete values in the union.
+    // invariant(abstractValue.isTemporal());
```
Why is this commented out?

trueadm commented:
Is this PR still active? Seems you addressed most of these issues in other PRs?
